core/CMakeLists.txt

```diff
@@ -103,6 +103,13 @@ target_compile_definitions( vecmem_core PUBLIC
    $<BUILD_INTERFACE:VECMEM_DEBUG_MSG_LVL=${VECMEM_DEBUG_MSG_LVL}>
    $<BUILD_INTERFACE:VECMEM_SOURCE_DIR_LENGTH=${VECMEM_SOURCE_DIR_LENGTH}> )
 
+# The library headers make checks on the value of the __cplusplus macro. So we
+# need to make sure that MSVC would always set that macro up when using this
+# library.
+target_compile_options( vecmem_core PUBLIC
+   $<$<AND:$<CXX_COMPILER_ID:MSVC>,$<COMPILE_LANGUAGE:CXX>>:/Zc:__cplusplus>
+   $<$<AND:$<CXX_COMPILER_ID:MSVC>,$<COMPILE_LANGUAGE:CUDA>>:-Xcompiler /Zc:__cplusplus> )
+
```

Comment:
I'm a little confused here, I was under the impression that `__cplusplus` was a C++-standard definition that should _always_ be there if compiling C++. Is this not the case? :fearful: Does MVCC not adhere to this standard?

Reply:
This is a messy thing that you're shining light on with this question... :stuck_out_tongue:

https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus

Some of the Windows headers don't like it when `__cplusplus` is set to a "too high value". :confused: So by default MSVC sets it to `199711L`. Whatever the actual C\+\+ standard is that it really uses.

So yeah, this is something quite unfortunate. But since we depend on the "correct" value of `__cplusplus` in our public header, we have to make sure in this project's code that MSVC would always set that macro value correctly...

Reply:
My god that is horrendously stupid (on MSVC's end). But yeah, I guess we need to fix it in this way then.

How on earth does anyone use MSVC to compile anything in production??